doc/manuals/sprokit/distributed_pipeline.rst

```diff
@@ -1,50 +1,58 @@
-- Describe how the port names work
-- Describe how the serialization algo is selected
-- Describe requirements for item names and types
-- Describe the transport process (notionally?)
-
-
-How to create a new data_serializer algorithm
-----------------------------------------------
-
-1) describe interface (base class)
+======================
+
+
+A key component required in  KWIVER to enable the
+construction of fully elaborated computer vision systems is a strategy for
+multi-processing by distributing Sprokit pipelines across multiple computing
+nodes.   This a critical requirement since modern computer vision algrorithms
+tend to be resource hungry, especially in the case of deep learning based
+alrorithms which require extensive GPU support to run optimally.  KWIVER has
+been utilized in systems using the
+`The Robotics Operating System (ROS) <http://www.ros.org>>`_ and `Apache
+Kafka <https://kafka.apache.org/>`_ among others.
+
+KWIVER, howver,  can use a built-in mechanism for constructing multi-computer
+processing systems, with message passing becoming an integral part of the
+KWIVER framework. We have chosen to use `ZeroMQ <https://zeromq.org>`_ for the
+message passing architecture, because it readily scales from small brokerless
+prototype systems to more complex broker based architectures that span many
+dozens of communicating elements.
+
+The current ZeroMQ system focuses on “brokerless” processing, relying on pipeline
+configuration settings to establish communication topologies. What this means
+in practice is that pipelines must be constructed in a way that “knows” where
+their communication partners are located in terms of networking (hostname and
+ports). While this is sufficient to stand up a number of interesting and useful
+systems  it is expected that KWIVER will evolve to providing limited brokering
+services to enable more flexibility and dynamism when constructing KWIVER based
+multi-processing systems.
+
+KWIVER's multi-processing support is composed of two components:
+
+#. Serialiation
+#. Transport
+
+In keeping with KWIVER's architecture, both of these are represented as
+abstractions, under which specific implementations (JSON, Protocol Buffers,
+ZeroMQ, ROS etc.) can be constructed.
+
+
+Serialization
+-------------
+
+KWIVER’s serialization strategy is based on KWIVER’s arrows. There is a
+serialization arrow for each of the VITAL data types. Then there are
+implementations for various serialization protocols.  KWIVER supports JSON
+based serialization and binary serialization. For binary serialization,
+Google’s Protocol Buffers are used. While JSON based serialization makes
+reading and debugging types like detected_object_set easy, binary serialization
+is used to serialize data heavy elements like images.    As with other KWIVER
+arrows, providing new implementations supporting other protocols is
+straightforward.
+
+Constructing a Serialization Algorithm
+''''''''''''''''''''''''''''''''''''''
```

Comment:
Sphinx section guideline puts ``^^^`` as the next section tier ([reference](http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections))